mlflow/server/auth/__init__.py

```diff
@@ -9,24 +9,269 @@
 
 import logging
 
+from flask import Flask, request, make_response, Response, redirect, flash
+
+from mlflow import get_run
 from mlflow.server import app
+from mlflow.server.auth import store
+from mlflow.server.auth.config import app_config
+from mlflow.server.auth.permissions import Permission, get_permission
+from mlflow.server.handlers import (
+    _get_tracking_store,
+    _get_request_message,
+    get_endpoints,
+    message_to_json,
+)
+from mlflow.protos.service_pb2 import (
+    CreateExperiment,
+    GetExperiment,
+    GetRun,
+    SearchRuns,
+    ListArtifacts,
+    GetMetricHistory,
+    CreateRun,
+    UpdateRun,
+    LogMetric,
+    LogParam,
+    SetTag,
+    SearchExperiments,
+    DeleteExperiment,
+    RestoreExperiment,
+    RestoreRun,
+    DeleteRun,
+    UpdateExperiment,
+    LogBatch,
+    DeleteTag,
+    SetExperimentTag,
+    GetExperimentByName,
+    LogModel,
+)
 
 
 _logger = logging.getLogger(__name__)
 
 
+class ROUTES:
+    CREATE_EXPERIMENT_PERMISSION = "/mlflow/experiments/permissions/create"
+    GET_EXPERIMENT_PERMISSION = "/mlflow/experiments/permissions/get"
+    UPDATE_EXPERIMENT_PERMISSION = "/mlflow/experiments/permissions/update"
+    DELETE_EXPERIMENT_PERMISSION = "/mlflow/experiments/permissions/delete"
+    USERS = "/users"
+    SIGNUP = "/signup"
+
+
+UNPROTECTED_ROUTES = [ROUTES.SIGNUP, ROUTES.USERS]
+
+
+def is_unprotected_route(path: str) -> bool:
+    if path.startswith("/static-files"):
+        return True
+    return path in UNPROTECTED_ROUTES
+
+
+def make_forbidden_response() -> Response:
+    res = make_response("Permission denied")
+    res.status_code = 403
+    return res
+
+
+def make_invalid_login_response() -> Response:
+    res = make_response("Invalid username or password")
+    res.status_code = 403
+    return res
+
+
+def make_basic_auth_response() -> Response:
+    res = make_response()
+    res.status_code = 401
+    res.headers["WWW-Authenticate"] = 'Basic realm="mlflow"'
+    return res
```

Comment:
@gabrielfu Thanks for the PR! Can we split up the authetication part into a separate PR? We can tackle that first :)